mpdev/internal/tf/overwrite.go

```diff
@@ -1,8 +1,142 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
 package tf
 
-import ()
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/hashicorp/hcl/v2"
+	"github.com/hashicorp/hcl/v2/hclsyntax"
+	"github.com/hashicorp/hcl/v2/hclwrite"
+	"github.com/hashicorp/terraform-config-inspect/tfconfig"
+	"os"
+)
+
+type overwriteConfig struct {
+	Variables    []string
+	Replacements map[string]string
+}
 
 // Overwrite replaces variable values in Terraform modules
-func Overwrite() error {
+func Overwrite(configBytes []byte, dir string) error {
+	config, err := getOverwriteConfig(configBytes)
+
+	if err != nil {
+		return fmt.Errorf("failure parsing overwrite config: %s error: %w", string(configBytes), err)
+	}
+	fmt.Printf("Replacing the default values of the variables: %s\n", config.Variables)
+	fmt.Printf("Mapping of values to replace: %s\n", config.Replacements)
+
+	for _, varname := range config.Variables {
+		varInfo, err := getVarInfo(varname, dir)
+		if err != nil {
+			return err
+		}
+
+		if varInfo.Default == nil {
+			return fmt.Errorf("image variable: %s must have default value", varname)
+		}
+
+		defaultVal, ok := varInfo.Default.(string)
+		if !ok {
+			return fmt.Errorf("image variable: %s must be type string", varname)
+		}
+
+		replaceVal, ok := config.Replacements[defaultVal]
+		if !ok {
+			return fmt.Errorf("default value: %s of variable: %s not found in replacements",
+				defaultVal, varname)
+		}
+
+		err = overwriteFile(varInfo.Pos.Filename, varname, replaceVal)
+		if err != nil {
+			return err
+		}
+	}
 	return nil
 }
+
+func getOverwriteConfig(b []byte) (*overwriteConfig, error) {
+	var config overwriteConfig
+	err := json.Unmarshal(b, &config)
+	if err != nil {
+		return nil, err
+	}
+
+	return &config, nil
+}
+
+func getVarInfo(varname string, dir string) (*tfconfig.Variable, error) {
+	module, diag := tfconfig.LoadModule(dir)
+	if diag.HasErrors() {
+		return nil, fmt.Errorf("failure parsing terraform module: %w", diag)
+	}
+
+	variable, ok := module.Variables[varname]
+	if !ok {
+		return nil, fmt.Errorf("variable: %s not found in module", varname)
+	}
+
+	return variable, nil
+}
+
+func getAttributeValueTokens(value string) hclwrite.Tokens {
+	return hclwrite.Tokens{
+		{
+			Type:         hclsyntax.TokenOQuote,
+			Bytes:        []byte(`"`),
+			SpacesBefore: 1,
+		},
+		{
+			Type:  hclsyntax.TokenQuotedLit,
+			Bytes: []byte(value),
+		},
+		{
+			Type:  hclsyntax.TokenOQuote,
```

Comment:
Should this be a TokenCQuote?

Reply:
Nice catch! I added a comment also to reference the HCL implementation which we should be following